### ISSUE TITLE
Lots of THROWIFNOT errors in mainnet

### ISSUE THREAD

**Issue by roman-khimov** [nspcc-dev/neo-go#454]
```
WARN[4356] contract invocation failed                    block=1953447 err="error encountered at instruction 80 (THROWIFNOT): THROWIFNOT" tx=8e232fd569bd2bd3f54bc45688f7496482fd059a09725ef860203d0af650fd0f
WARN[4356] contract invocation failed                    block=1953448 err="error encountered at instruction 80 (THROWIFNOT): THROWIFNOT" tx=1c78818f72839011aa0a3d0a08ab4f8b02ec6e47e0f48a1163090334b2f0e145
WARN[4356] contract invocation failed                    block=1953449 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=4108a67687d0463abba7b6b17bfb3c3e567b25aad2ddd2fe1db5670bffdd35a7
```
Similar to #453, there are lots of these, so it's probably a VM issue too.

**Comment by roman-khimov** [comment]
First one to fail this way:
```
INFO[1998] blockchain persist completed                  blockHeight=1444903 headerHeight=1444904 persistedBlocks=1 persistedKeys=13337 took=32.136484ms
INFO[1999] blockchain persist completed                  blockHeight=1444905 headerHeight=1444906 persistedBlocks=2 persistedKeys=4526 took=8.337866ms
WARN[1999] contract invocation failed                    block=1444910 err="error encountered at instruction 80 (THROWIFNOT): THROWIFNOT" tx=684e1b6e45e54b843a2445e0a15f9734db5c126dcb07092dc9a30b34e08da086
INFO[2000] blockchain persist completed                  blockHeight=1444912 headerHeight=1444913 persistedBlocks=7 persistedKeys=4944 took=13.440174ms
INFO[2001] blockchain persist completed                  blockHeight=1444917 headerHeight=1444918 persistedBlocks=5 persistedKeys=4150 took=8.559236ms
INFO[2002] blockchain persist completed                  blockHeight=1444923 headerHeight=1444924 persistedBlocks=6 persistedKeys=4181 took=11.22011ms
INFO[2003] blockchain persist completed                  blockHeight=1444931 headerHeight=1444932 persistedBlocks=8 persistedKeys=5458 took=7.275327ms
INFO[2004] blockchain persist completed                  blockHeight=1444940 headerHeight=1444941 persistedBlocks=9 persistedKeys=5533 took=10.801393ms
WARN[2004] contract invocation failed                    block=1444942 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=1bcfb5a4813985235ea0f5d68d38ae646c1a4b820f2e487fdb7b11a6a50cf9b0
WARN[2004] contract invocation failed                    block=1444943 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=47b3a86fb482340da5b976cd588865d5111a15139c9a3b296dfd675f4f57fdd3
WARN[2004] contract invocation failed                    block=1444944 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=16f55c23bf15de61b3ea3674544ff83db5a1899bad8b504c8922c0a5a58a2db0
INFO[2005] blockchain persist completed                  blockHeight=1444946 headerHeight=1444947 persistedBlocks=6 persistedKeys=4655 took=8.239206ms
WARN[2005] contract invocation failed                    block=1444950 err="error encountered at instruction 79 (THROWIFNOT): THROWIFNOT" tx=3f4031a0106856b19146df52184aa78f6cb43b03e11eb397d798031fef009cc4
INFO[2006] blockchain persist completed                  blockHeight=1444952 headerHeight=1444953 persistedBlocks=6 persistedKeys=5052 took=10.729689ms
WARN[2006] contract invocation failed                    block=1444955 err="error encountered at instruction 79 (THROWIFNOT): THROWIFNOT" tx=12150f0f6d86f4171603ab48bfe4d2d2ff315ce684c282ab5e255926587f6718
INFO[2007] blockchain persist completed                  blockHeight=1444961 headerHeight=1444962 persistedBlocks=9 persistedKeys=4807 took=12.968531ms
WARN[2008] contract invocation failed                    block=1444969 err="error encountered at instruction 79 (THROWIFNOT): THROWIFNOT" tx=76e7fcc927c841c68b817cdfb8f43dbb8591cc01616d1541c780bae6e7c5a075
INFO[2008] blockchain persist completed                  blockHeight=1444971 headerHeight=1444972 persistedBlocks=10 persistedKeys=5464 took=11.367505ms
INFO[2009] blockchain persist completed                  blockHeight=1444987 headerHeight=1444988 persistedBlocks=16 persistedKeys=6223 took=10.252118ms
WARN[2009] contract invocation failed                    block=1444991 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=1100a104a5912502f76030a880c16ee9a31f8f744c3321b2b6cb01069881d9d5
WARN[2009] contract invocation failed                    block=1444997 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=00149123c1a8f586ea61ff870cee03e2fb93491d64882ea8ed08b8169424cfc9
WARN[2009] contract invocation failed                    block=1445003 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=33daa2a69f0c90cae247ff410bc735b12ddb2a2cb73d7b1c022f521da12a7071
WARN[2010] contract invocation failed                    block=1445008 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=0c1ac282bd7a3fd0e4b268403d6b1429c2344363e01517121152a992b9602ce9
WARN[2010] contract invocation failed                    block=1445009 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=e68f4b78cc34d293f7eb6612b88a903364d0164f9437e8fc5eb5c53c598243d1
INFO[2010] blockchain persist completed                  blockHeight=1445012 headerHeight=1445013 persistedBlocks=25 persistedKeys=6794 took=18.349751ms
INFO[2011] blockchain persist completed                  blockHeight=1445024 headerHeight=1445025 persistedBlocks=12 persistedKeys=2973 took=5.580666ms
```

**Comment by roman-khimov** [comment]
Stack contents on `THROWIFNOT`:
```
[
    {
        "value": "",
        "type": "ByteArray"
    }
]
```

**Comment by roman-khimov** [comment]
Contract that fails: [ecc6b20d3ccac1ee9ef109af5a7cdb85706b1df9.avm.gz](https://github.com/nspcc-dev/neo-go/files/3841838/ecc6b20d3ccac1ee9ef109af5a7cdb85706b1df9.avm.gz).

Tx link on neoscan: https://neoscan.io/transaction/684E1B6E45E54B843A2445E0A15F9734DB5C126DCB07092DC9A30B34E08DA086

Minting tx for the token: https://neoscan.io/transaction/2471506a06d78e9c4ad12e4535f3b79bcec889fdec1561d33043858560fbdde5

**Comment by roman-khimov** [comment]
Something's really wrong here. On second run I've got this one passing, but now block 1444907 failed on tx 246d8fd5195b2fa11602732939245fd7808b8df087033fbb0ad62876dd47ec67.

**Comment by roman-khimov** [comment]
```
WARN[0003] contract invocation failed                    block=1444913 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=0ba3f261886a75f1f3d486bd4a1f6b6a0809b5c6a533c62f816e4ff125159e54
```
So it's kinda random.

**Comment by roman-khimov** [comment]
Another run (BoltDB backend now):
```
time="2019-11-13T20:27:54+03:00" level=info msg="blockchain persist completed" blockHeight=1444933 headerHeight=1444934 persistedBlocks=12 persistedKeys=7349 took=138.085968ms
time="2019-11-13T20:27:55+03:00" level=warning msg="contract invocation failed" block=1444939 err="error encountered at instruction 80 (THROWIFNOT): THROWIFNOT" tx=990e188ab5db18f3012e3ff148c69794d7e6153821db0854440a1a9c8fd97291
time="2019-11-13T20:27:55+03:00" level=warning msg="contract invocation failed" block=1444943 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=47b3a86fb482340da5b976cd588865d5111a15139c9a3b296dfd675f4f57fdd3
time="2019-11-13T20:27:55+03:00" level=info msg="blockchain persist completed" blockHeight=1444944 headerHeight=1444945 persistedBlocks=11 persistedKeys=6785 took=128.125314ms
time="2019-11-13T20:27:56+03:00" level=warning msg="contract invocation failed" block=1444947 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=94d6650f04a4201de7f97cc0ac6d9005a8d65df1deb01f03a5417497e4b0b4aa
time="2019-11-13T20:27:56+03:00" level=warning msg="contract invocation failed" block=1444952 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=306a4dca1a8fbbfe9494827a037d1ee21157e91a013c6bc09a8ef28d3676ca85
time="2019-11-13T20:27:56+03:00" level=info msg="blockchain persist completed" blockHeight=1444951 headerHeight=1444952 persistedBlocks=7 persistedKeys=6677 took=123.915058ms
time="2019-11-13T20:27:56+03:00" level=warning msg="contract invocation failed" block=1444953 err="error encountered at instruction 79 (THROWIFNOT): THROWIFNOT" tx=5e5b9def99c9fa19c51875e7f13bcc79950f0fa8907eec4182611df8233912ce
time="2019-11-13T20:27:57+03:00" level=warning msg="contract invocation failed" block=1444961 err="error encountered at instruction 79 (THROWIFNOT): THROWIFNOT" tx=eabce092cbfa8e0cf1d55b9c7cfd9a3aa93a7f19d30b70a1314ca6ba57aec824
time="2019-11-13T20:27:57+03:00" level=info msg="blockchain persist completed" blockHeight=1444964 headerHeight=1444965 persistedBlocks=13 persistedKeys=7215 took=129.11972ms
time="2019-11-13T20:27:58+03:00" level=info msg="blockchain persist completed" blockHeight=1444983 headerHeight=1444984 persistedBlocks=19 persistedKeys=8238 took=146.163697ms
time="2019-11-13T20:27:59+03:00" level=warning msg="contract invocation failed" block=1445001 err="error encountered at instruction 80 (THROWIFNOT): THROWIFNOT" tx=1e927956a5ff216a3aa955f794f7f555ca6a253332ac26e7f225a1e4646f493b
time="2019-11-13T20:27:59+03:00" level=warning msg="contract invocation failed" block=1445005 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=e8584690365249af65910fcb759abd837cdc99f13d2da6798e29bfc5ced8fa21
time="2019-11-13T20:27:59+03:00" level=warning msg="contract invocation failed" block=1445009 err="error encountered at instruction 81 (THROWIFNOT): THROWIFNOT" tx=e68f4b78cc34d293f7eb6612b88a903364d0164f9437e8fc5eb5c53c598243d1
time="2019-11-13T20:27:59+03:00" level=info msg="blockchain persist completed" blockHeight=1445021 headerHeight=1445021 persistedBlocks=38 persistedKeys=10090 took=172.627635ms
time="2019-11-13T20:28:00+03:00" level=info msg="blockchain persist completed" blockHeight=1445024 headerHeight=1445025 persistedBlocks=3 persistedKeys=935 took=35.346725ms
time="2019-11-13T20:28:04+03:00" level=info msg="blockchain persist completed" blockHeight=1445025 headerHeight=1445026 persistedBlocks=1 persistedKeys=25300 took=531.084365ms
```

**Comment by roman-khimov** [comment]
So we have some nondeterminism here. I thought that it might be related to persistence logic, but a quick mutexing attempt shows that isolating `AddBlocks` and `persist` doesn't help. As that's the only thing happening here (the main goroutine adds blocks and the one from `Run` periodically persist the DB) it most probably is related to the GC/and or slice ownerships. But at the same time on the DB level these issues were fixed a long time ago (ac5d2f94d3413143a8d0a46742de8ce9eeeb1b2c or 13bf2618effbd2ef35a71204628b443d85b92750) so I'm not sure.

**Comment by roman-khimov** [comment]
The problem occurs in the previous transaction that should had put something in the DB, no `Neo.Storage.Put` registered for failing subsequent transfers.
